sympy/polys/series/powerseriesring.py

```diff
@@ -0,0 +1,149 @@
+from __future__ import annotations
+
+from typing import Any, Protocol, TypeVar
+from sympy.core.symbol import Symbol
+from sympy.polys.domains import Domain
+
+
+_x: Symbol = Symbol("x")
+TSeries = TypeVar("TSeries")
+
+
+def _series_from_list(series: list[Any], prec: int | None) -> str:
+    """Convert a list of coefficients into a string representation of a power series."""
+    terms = []
+
+    if prec == 0:
+        return f"O({_x}**{prec})"
+
+    for i, coeff in enumerate(series):
+        if coeff == 0:
+            continue
+        if coeff == 1 and i != 0:
+            term = f"{_x}" if i == 1 else f"{_x}**{i}"
+        elif coeff == -1 and i != 0:
+            term = f"-{_x}" if i == 1 else f"-{_x}**{i}"
+        else:
+            if i == 0:
+                term = f"{coeff}"
+            elif i == 1:
+                term = f"{coeff}*{_x}"
+            else:
+                term = f"{coeff}*{_x}**{i}"
+        terms.append((i, term))
+
+    if not terms:
+        if prec is not None:
+            return f"0 + O({_x}**{prec})"
+        else:
+            return "0"
+
+    poly = " + ".join(term for _, term in terms).replace("+ -", "- ")
+    if prec is not None:
+        return f"{poly} + O({_x}**{prec})"
+    return poly
+
+
+class PowerSeriesRing(Protocol[TSeries]):
+    """Protocol for power series ring over the integer ring (ZZ).
```

Comment:
I'm not sure about this but I think it usually makes sense to include `_proto` or something in the name of a Protocol so that it is clear to calling code that it is not a runtime type.